editor/src/messages/layout/layout_message_handler.rs

```diff
@@ -208,55 +239,57 @@ impl<F: Fn(&MessageDiscriminant) -> Vec<KeysGroup>> MessageHandler<LayoutMessage
 }
 
 impl LayoutMessageHandler {
+	/// Diff the update and send to the frontend where necessary
+	fn send_layout(&mut self, layout_target: LayoutTarget, new_layout: Layout, responses: &mut VecDeque<Message>, action_input_mapping: &impl Fn(&MessageDiscriminant) -> Vec<KeysGroup>) {
+		// We don't diff the menu bar layout yet.
+		if layout_target == LayoutTarget::MenuBar {
+			// Skip update if the same
+			if self.layouts[layout_target as usize] == new_layout {
+				return;
+			}
+			// Update the backend storage
+			self.layouts[layout_target as usize] = new_layout;
+			// Update the UI
+			responses.push_back(
+				FrontendMessage::UpdateMenuBarLayout {
+					layout_target,
+					layout: self.layouts[layout_target as usize].clone().unwrap_menu_layout(action_input_mapping).layout,
+				}
+				.into(),
+			);
+			return;
+		}
+
+		let mut widget_diffs = Vec::new();
```

Comment:
I think that we should only do diffs for when we `RefreshLayout`. I think we need a hybrid system where a diff or the entire layout can be sent. Every time something on the backend sends a layout all of the `widget_ids` are changed. I think we can do something to compare shapes and update on fine-grained details. However, I think that this is out of scope for just the initial work

Reply:
> I think we need a hybrid system where a diff or the entire layout can be sent. 

We currently send always the diff and there are no issues? Or am I misunderstanding?

Reply:
I think always sending the diff is less efficient in many cases than what we currently are doing. This is because even if nothing changes in the layout all of the `widget_ids` change. I think the system would be simpler if diffs were only of individual widgets.

Reply:
To add, it does not seem like this affected the memory usage of UI elements in preliminary testing on the preview

Reply:
`RefreshLayout` I renamed to `ResendActiveWidget` as that is what it does now.

Reply:
> Every time something on the backend sends a layout all of the widget_ids are changed.

No - a new layout is generated with new ids but we don't actually change the ids of the cached layout unless part of it has changed.

Reply:
> This is because even if nothing changes in the layout all of the widget_ids change.

Well a new layout is generated with new ids, but they do not get updated in the cached layout or in the frontend so same on update time unless the widget has actually changed.

Reply:
> To add, it does not seem like this affected the memory usage of UI elements in preliminary testing on the preview

This was not intended to reduce memory usage, only to reduce the amount of time per frame serialising data with serde.